### PR TITLE
Copy 'Ready' condition from VMI to VM

### LINKED ISSUE #2864: Provide Ready condition for VM
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind enhancement

**What happened**:
$ virtctl start testvm
$ kubectl wait --for=condition=Ready vm testvm
No such condition

The condition is just not available today, only a phsae.

**What you expected to happen**:
$ virtctl start testvm
$ kubectl wait --for=condition=Ready vm testvm
# vm running
$ virtctl vnc testvm
# vnc can connect

**How to reproduce it (as minimally and precisely as possible)**:
Just run the lines above

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`):
- Kubernetes version (use `kubectl version`):
- VM or VMI specifications:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by akrejcir**
When 'Ready' condition is added the VMI resource,
the virt-controller copies it to the VM resource.

Signed-off-by: Andrej Krejcir <akrejcir@redhat.com>

**What this PR does / why we need it**:
Users will be able to wait for the VM start using `kubectl wait --for=condition=Ready vm ${NAME}`

**Which issue(s) this PR fixes**
Fixes #2864

**Release note**:
```release-note
Added a new 'Ready' condition to the VirtualMachine resource.
The condition is copied from the associated VMI resource.

This makes it possible to wait while the VM starts on the VM resource:
$ kubectl wait --for=condition=Ready vm ${NAME}
```

### CODE REVIEW

File: pkg/virt-controller/watch/vm.go [hunk 1]
```diff
@@ -1285,6 +1294,38 @@ func (c *VMController) updateStatus(vmOrig *virtv1.VirtualMachine, vmi *virtv1.V
 	return err
 }
 
+func (c *VMController) copyReadyConditionFromVMI(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) {
```
Comment (dhiller) [inline pkg/virt-controller/watch/vm.go]: nit: rather call it `syncReadyConditionFromVMI`?

Comment (akrejcir) [reply pkg/virt-controller/watch/vm.go]: Done

File: pkg/virt-controller/watch/vm.go [hunk 2]
```diff
@@ -1285,6 +1294,38 @@ func (c *VMController) updateStatus(vmOrig *virtv1.VirtualMachine, vmi *virtv1.V
...
+			vm.Status.Conditions = append(vm.Status.Conditions, newCond)
+		} else if vmiReadyCond.Status != vmReadyCond.Status {
+			log.Log.Object(vm).V(4).Info("Updating ready condition")
+			vmReadyCond.Status = vmiReadyCond.Status
+			vmReadyCond.LastTransitionTime = now
+			vmReadyCond.Reason = vmiReadyCond.Reason
+			vmReadyCond.Message = vmiReadyCond.Message
+
+		}
+	} else if vmReadyCond != nil {
```
Comment (danielBelenky) [inline pkg/virt-controller/watch/vm.go]: This edge case is redundant. I think that it will also improve the readability of this function:

```
if vmiReadyCond == nil {
    c.removeCondition(vm, virtv1.VirtualMachineReady)
    return
}

// now just copy the condition
```

File: pkg/virt-controller/watch/vm.go [hunk 3]
```diff
@@ -1285,6 +1294,38 @@ func (c *VMController) updateStatus(vmOrig *virtv1.VirtualMachine, vmi *virtv1.V
...
+				Type:               virtv1.VirtualMachineReady,
+				Status:             vmiReadyCond.Status,
+				LastProbeTime:      now,
+				LastTransitionTime: now,
+				Reason:             vmiReadyCond.Reason,
+				Message:            vmiReadyCond.Message,
+			}
+			vm.Status.Conditions = append(vm.Status.Conditions, newCond)
+		} else if vmiReadyCond.Status != vmReadyCond.Status {
+			log.Log.Object(vm).V(4).Info("Updating ready condition")
```
Comment (danielBelenky) [inline pkg/virt-controller/watch/vm.go]: Why not always update the VM ready condition? regardless of it's value...
If we don't update it, the timestamps (at least the `LastProbeTime` will become stale and it usually indicates of an issue to the user).

File: pkg/virt-controller/watch/vm_test.go [hunk 4]
```diff
@@ -818,6 +818,40 @@ var _ = Describe("VirtualMachine", func() {
...
+				objVM := obj.(*v1.VirtualMachine)
+				cond := getVmCondition(objVM, v1.VirtualMachineReady)
+				Expect(cond).ToNot(BeNil())
+				Expect(cond.Status).To(Equal(k8sv1.ConditionTrue))
+			}).Return(vm, nil)
+
+			controller.Execute()
+		})
+
+		It("should add ready condition False", func() {
```
Comment (danielBelenky) [inline pkg/virt-controller/watch/vm_test.go]: Can those tests become a table? since the flow is similar and they differ only in the value of the condition...

File: pkg/virt-controller/watch/vm_test.go [hunk 5]
```diff
@@ -1011,3 +1045,15 @@ func DefaultVirtualMachineWithNames(started bool, vmName string, vmiName string)
 func DefaultVirtualMachine(started bool) (*v1.VirtualMachine, *v1.VirtualMachineInstance) {
 	return DefaultVirtualMachineWithNames(started, "testvmi", "testvmi")
 }
+
+func getVmCondition(vm *virtv1.VirtualMachine, condType virtv1.VirtualMachineConditionType) *virtv1.VirtualMachineCondition {
```
Comment (danielBelenky) [inline pkg/virt-controller/watch/vm_test.go]: why is it needed? can't we use `getCondition`? https://github.com/kubevirt/kubevirt/pull/3176/files#diff-3d28bf4913d346c1909faf1a6411f176R48

File: tests/vm_test.go [hunk 6]
```diff
@@ -634,6 +634,31 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
 			}
 		})
 
+		It("should set the Ready condition on VM", func() {
+			vm := newVirtualMachine(false)
+
+			getVmReadyConditionStatus := func() *k8sv1.ConditionStatus {
```
Comment (danielBelenky) [inline tests/vm_test.go]: Same here... why can't we use `getCondition` and then check the Status?

File: pkg/virt-controller/watch/vm.go (hunk 3 above)
Comment (akrejcir) [reply pkg/virt-controller/watch/vm.go]: I wanted to minimize update requests, but this makes sense. Should the `LastProbeTime` be copied from VMI or set to current time?

Comment (danielBelenky) [reply pkg/virt-controller/watch/vm.go]: I think that copying `LastProbeTime` makes sense...

File: pkg/virt-controller/watch/vm.go (hunk 2 above)
Comment (akrejcir) [reply pkg/virt-controller/watch/vm.go]: That would call the `c.removeCondition` function on every update even when it is not needed. I would prefer to leave the condition there, but I have made it more readable.

File: pkg/virt-controller/watch/vm.go (hunk 3 above)
Comment (akrejcir) [reply pkg/virt-controller/watch/vm.go]: Done

File: pkg/virt-controller/watch/vm_test.go (hunk 4 above)
Comment (akrejcir) [reply pkg/virt-controller/watch/vm_test.go]: Done

File: pkg/virt-controller/watch/vm_test.go (hunk 5 above)
Comment (akrejcir) [reply pkg/virt-controller/watch/vm_test.go]: Moved the function to a common place `pkg/controller/conditions.go` and it is called when needed.

File: tests/vm_test.go (hunk 6 above)
Comment (akrejcir) [reply tests/vm_test.go]: Done

File: pkg/controller/conditions.go [hunk 7]
```diff
@@ -25,10 +25,44 @@ import (
 	v1 "kubevirt.io/client-go/api/v1"
 )
 
-type VirtualMachineConditionManager struct {
+type virtualMachineConditionManager struct {
 }
 
-func (d *VirtualMachineConditionManager) CheckFailure(vmi *v1.VirtualMachineInstance, syncErr error, reason string) (changed bool) {
```
Comment (danielBelenky) [inline pkg/controller/conditions.go]: Why you changed this? it used to be exported and now it is not...

File: pkg/controller/conditions.go [hunk 8]
```diff
@@ -45,31 +79,28 @@ func (d *VirtualMachineConditionManager) CheckFailure(vmi *v1.VirtualMachineInst
 	return false
 }
 
-func (d *VirtualMachineConditionManager) HasCondition(vmi *v1.VirtualMachineInstance, cond v1.VirtualMachineInstanceConditionType) bool {
+func (d virtualMachineInstanceConditionManager) GetCondition(vmi *v1.VirtualMachineInstance, cond v1.VirtualMachineInstanceConditionType) *v1.VirtualMachineInstanceCondition {
```
Comment (danielBelenky) [inline pkg/controller/conditions.go]: same here...

File: pkg/controller/conditions.go [hunk 9]
```diff
@@ -129,14 +160,14 @@ func (d *VirtualMachineConditionManager) GetPodCondition(pod *k8sv1.Pod, conditi
 	return nil
 }
 
-func NewVirtualMachineInstanceConditionManager() *VirtualMachineConditionManager {
-	return &VirtualMachineConditionManager{}
+func VmiConditionManager() virtualMachineInstanceConditionManager {
```
Comment (danielBelenky) [inline pkg/controller/conditions.go]: Is this change required for this PR?

File: pkg/virt-controller/watch/vm.go (hunk 2 above)
Comment (danielBelenky) [reply pkg/virt-controller/watch/vm.go]: I don't see a real harm in calling `c.removeCondition` every time (it will do nothing if the condition is nil)...
But it's your code so I don't mind that much.
Thanks for improving the readability :+1:

File: pkg/controller/conditions.go (hunk 9 above)
Comment (akrejcir) [reply pkg/controller/conditions.go]: No, not required. I thought the names were too long. It is in a separate commit, should it be a separate PR too?

File: pkg/controller/conditions.go (hunk 7 above)
Comment (akrejcir) [reply pkg/controller/conditions.go]: Does the struct need to be exported? It has no use other than calling it's method, and it can be accessed through the exported method.

File: pkg/controller/conditions.go (hunk 9 above)
Comment (danielBelenky) [reply pkg/controller/conditions.go]: I prefer it to go to a separate PR if possible since its unrelated.

File: pkg/controller/conditions.go (hunk 7 above)
Comment (danielBelenky) [reply pkg/controller/conditions.go]: If this struct is not used anywhere else, we can un-export it. But, since this is not related to this PR, maybe we can do it in a separate PR where we can discuss if it's required or not?

(I'm sorry for being picky on this...)

File: pkg/controller/conditions.go (hunk 9 above)
Comment (akrejcir) [reply pkg/controller/conditions.go]: Ok, the other commit is removed.

**Comment by akrejcir** [comment]
`/retest`

**Comment by danielBelenky** [comment]
Hey @akrejcir, thanks for addressing the comments.
I see that there are some test failures so once those are resolved, we can get this PR merged.

**Comment by akrejcir** [comment]
The travis CI is failing, but the failure does not seem related to this patch. Is there something I can do to fix it?

**Comment by rmohr** [comment]
I see that your PR is still based on a branch were  we run the PPC build which we have now temporarily disabled. Could you just rebase?

File: pkg/controller/conditions.go [hunk 10]
```diff
@@ -28,7 +28,41 @@ import (
...
+	for _, c := range vm.Status.Conditions {
+		if c.Type == cond {
+			continue
+		}
+		conds = append(conds, c)
+	}
+	vm.Status.Conditions = conds
+}
+
+type VirtualMachineInstanceConditionManager struct {
```
Comment (xpivarc) [inline pkg/controller/conditions.go:62]: Is there a need to have this as struct? Could be this only simple functions?

Comment (akrejcir) [reply pkg/controller/conditions.go:62]: I used the same design as the other structs in this file use, to be consistent. I agree that the structs are useless, we could remove them and rename all the functions in a new PR. Function overloading would be ideal for this use case, but Go does not support it.

**Comment by akrejcir** [comment]
The tests have passed, can this be merged?
